### ISSUE TITLE
[Bug]: Recursively calling liftListItem fails to lift the listItem out of its list without any clue to the problem

### ISSUE THREAD

**Issue by Slapbox** [ueberdosis/tiptap#4283]
### Which packages did you experience the bug in?

extension-list-item

### What Tiptap version are you using?

2.1.0-rc.12

### What’s the bug you are facing?

If you want to call `liftListItem` repeatedly until an item would no longer be in a list, it's not possible even though it should be. Instead the `listItem` stays at the shallowest depth in the list and it requires an entirely separate transaction to get it to finally leave the list.

**Specifically this issue occurs when you try to lift an item which you just created via splitting. For example: `editor.chain().splitListItem('listItem').liftListItem('listItem').liftListItem('listItem').run()`**

This blocks off all sorts of functionality related to easily transforming nodes from one type to another, which is seriously limiting and prevents us from fully porting our editor to TipTap currently.

### What browser are you using?

Chrome

### What did you expect to happen?

The newly split item should lift out of the list, not just to the top level of the list.

### Did you update your dependencies?

- [X] Yes, I’ve updated my dependencies to use the latest version of all packages.

### Are you sponsoring us?

- [ ] Yes, I’m a sponsor. 💖

**Comment by Slapbox** [comment]
Specifically this issue is caused by this line in ProseMirror: `if (tr.mapping.map(range.end) != range.start + $start.nodeAfter!.nodeSize) return false`

Is there some way we can help get to `tr.mapping` result fixed here?

**Comment by Slapbox** [comment]
Actually, this problem does not appear to affect ProseMirror's `splitListItem` so the problem must be with the custom TipTap implementation.

**Comment by Slapbox** [comment]
@bdbch pinging you since you're the "lists guy."

Do you have any idea what might cause this? If so I'd be happy to investigate further and submit a PR.

**Comment by bdbch** [comment]
I'd wish I was the lists guy, haha! 😂

Seems like another upstream issue coming from Prosemirror - but not quite sure right now. I think I'd have to redirect to prosemirror for this as we're using their underlying lift functionality.

Feel free to submit a PR - maybe it would make more sense to fix this in Prosemirror though if possible. I was thinking about upgrading our prosemirror versions anyway pretty soon.

**Comment by Slapbox** [comment]
Thanks very much for your reply @bdbch!

I see I made a typo. What I meant was the custom [`splitListItem` code](https://github.com/ueberdosis/tiptap/blob/develop/packages/core/src/commands/splitListItem.ts) is the problem. I can confirm that the ProseMirror version of the code does not exhibit this problem.

I see the ProseMirror version was actually used up until [this commit](https://github.com/ueberdosis/tiptap/commit/30398717671f79e3884f9fd48da566c1248ab944), so maybe @philippkuehn could provide some insight?

**Comment by bdbch** [comment]
I think it depends. If it's something introduced on our end (for example because of chaining) we can take a look into it if it's possible to fix it from our side.

I'm a bit stuck with upstream issues and we talk a lot about those currently because they take a lot of time without actually benefiting the core product (Prosemirror) in the end and takes away time we could directly work on Tiptap related things.

I'd say if the chaining is a problem here, we should take a look. Nothing marjin can do there I guess.

Could you explain how exactly chaining introduces issues here?

**Comment by Slapbox** [comment]
It seems to be caused by the chaining, but not due to any error in the chaining logic per se. If I run the same commands without chaining them, everything works fine - but with chaining the `mapping` results end up being incorrect in some way that I'm not really clear on, and the logic in some very niche situations ends up failing due to the mapping results. I don't know if it's that the mapping is really off in these cases, or that ProseMirror's internal commands simply cannot handle the results of chained commands because they were never tested under those conditions by marijnh.

For what it's worth most (may even all) of these edge cases seem to be specifically related to using chaining together with `prosemirror-schema-list`

**Comment by bdbch** [comment]
Yeah I think the mapping can go off the rails when chained when doing more complex things. I already wondered if we're doing something wrong while chaining (maybe we have to process the mapping on each chain step to keep it correct?)

Right now I'm a bit in the dark on that one.

**Comment by Slapbox** [comment]
Just updating the notes here: this issue may require changes similar to a change made in this PR: https://github.com/ueberdosis/tiptap/pull/4210#issuecomment-1672257958

**Comment by tamasarpad** [comment]
Without chaining it worked for me.
This doesn't lift out the item from the list for me (stuck at the first level):
                             editor.chain()
                                .liftListItem(parentType);
                                .liftListItem(parentType);
                                .liftListItem(parentType);
                                .liftListItem(parentType);
                                .liftListItem(parentType);
                                .liftListItem(parentType);
                                .run();

Removing chaining lifts it out:
                                editor.commands.liftListItem(parentType);
                                editor.commands.liftListItem(parentType);
                                editor.commands.liftListItem(parentType);
                                editor.commands.liftListItem(parentType);
                                editor.commands.liftListItem(parentType);
                                editor.commands.liftListItem(parentType);

parentType is either 'listItem' or 'taskItem' in my code
